File: meme_inator_back/apps/posts/domain/irepositories/ipost_vote_repository.py
# apps/posts/domain/irepositories/ipost_vote_repository.py
from abc import ABC, abstractmethod
from typing import Optional
from uuid import UUID

from apps.posts.domain.entities.post_vote_entity import PostVoteEntity, PostVoteTypeEnum
from core.results import Result


class IPostVoteRepository(ABC):
    """Interface for post vote repository."""
    
    @abstractmethod
    def vote_on_post(
        self, 
        post_public_id: UUID, 
        user_id: UUID, 
        vote_type: PostVoteTypeEnum
    ) -> Result[PostVoteEntity]:
        """Add or update a vote on a post."""
        pass
    
    @abstractmethod
    def remove_vote(
        self, 
        post_public_id: UUID, 
        user_id: UUID
    ) -> Result[bool]:
        """Remove a user's vote from a post."""
        pass
    
    @abstractmethod
    def get_user_vote(
        self, 
        post_public_id: UUID, 
        user_id: UUID
    ) -> Result[Optional[PostVoteEntity]]:
        """Get a user's vote on a specific post."""
        pass
    
    @abstractmethod
    def get_post_vote_stats(
        self, 
        post_public_id: UUID
    ) -> Result[dict]:
        """Get vote statistics for a post."""
        pass
    
    @abstractmethod
    def does_post_exist(self, post_public_id: UUID) -> bool:
        """Check if a post exists."""
        pass

File: meme_inator_front/lib/features/post/domain/repositories/ipost_vote_repository.dart
// apps/posts/domain/irepositories/ipost_vote_repository.dart
import 'package:meme_inator_front/core/results.dart';
import 'package:meme_inator_front/features/post/domain/entities/post_vote_entity.dart';
import 'package:meme_inator_front/features/post/domain/enums/post_vote_type.dart';
import 'package:uuid/uuid.dart';


abstract class IPostVoteRepository {
  Result<PostVoteEntity> voteOnPost(
    UuidValue postPublicId,
    UuidValue userId,
    PostVoteTypeEnum voteType,
  );

  Result<bool> removeVote(
    UuidValue postPublicId,
    UuidValue userId,
  );

  Result<PostVoteEntity?> getUserVote(
    UuidValue postPublicId,
    UuidValue userId,
  );

  Result<Map<String, int>> getPostVoteStats(
    UuidValue postPublicId,
  );

  bool doesPostExist(UuidValue postPublicId);
}
